src/managers/GuildEmojiManager.js

```diff
@@ -45,19 +45,22 @@ class GuildEmojiManager extends BaseGuildEmojiManager {
    */
   async create(attachment, name, { roles, reason } = {}) {
     attachment = await DataResolver.resolveImage(attachment);
-    if (!attachment) throw new TypeError('REQ_RESOURCE_TYPE');
+    if (!attachment) return Promise.reject(new TypeError('REQ_RESOURCE_TYPE'));
 
     const data = { image: attachment, name };
     if (roles) {
+      if (!Array.isArray(roles) && !(roles instanceof Collection)) {
+        return Promise.reject(
+          new TypeError('INVALID_TYPE', 'options.roles', 'Array or Collection of Roles or Snowflakes', true),
+        );
+      }
       data.roles = [];
-      for (let role of roles instanceof Collection ? roles.values() : roles) {
-        role = this.guild.roles.resolve(role);
-        if (!role) {
-          return Promise.reject(
-            new TypeError('INVALID_TYPE', 'options.roles', 'Array or Collection of Roles or Snowflakes', true),
-          );
+      for (const role of roles instanceof Collection ? roles.values() : roles) {
+        const resolvedRole = this.guild.roles.resolve(role);
+        if (!resolvedRole) {
+          return Promise.reject(new TypeError('INVALID_ELEMENT', 'Array or Collection', 'options.roles', role));
         }
```

Comment:
```suggestion
        const resolvedRole = this.guild.roles.resolveID(role);
        if (!resolvedRole) throw new TypeError('INVALID_ELEMENT', 'Array or Collection', 'options.roles', role));
```
While `resolveID` makes any string valid, it allows users to pass ids of uncached roles

Reply:
I'd be happy to change this. I didn't because of exactly what you mentioned here. Discord will error if the role id is invalid, but I wasn't sure I should change these as roles are always expected to be cached (at least for now).
I suppose for those using typescript this isn't an issue due to the recently increased type safety of snowflakes.
Any further thoughts?